File: LeetCode/cpp/generateParenthesis.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <queue>

using namespace std;

class Solution {
public:
    void backtrack(vector<string>& v, int n, int left, int right, string s)
    {
        if(s.size() == 2*n)
        {
            v.push_back(s);
            return;
        }
        
        if(left < n)
            backtrack(v, n, left+1, right, s+'(');
        
        if(right < left)
            backtrack(v, n, left, right+1, s+')');
    }
    
    vector<string> generateParenthesis(int n) {
        vector<string> v;
        string s;
        backtrack(v, n, 0, 0, s);
        return v;
    }
};


int main()
{
    int n = 4;
    Solution s;
    vector<string> v = s.generateParenthesis(n);
    for(auto&x: v)
        cout << x << endl;
    return 0;
}

File: LeetCode/python/generateParenthesis.py
"""
给定n, 求能生成多少有效括号

例如n = 4
res = [
    ((()))
    (()())
    (())()
    ()(())
    ()()()
]
"""


def backtrack(res, s, n, left, right):
    if len(s) == 2 * n:
        res.append(s)

    if left < n:
        backtrack(res, s+"(", n, left+1, right)

    if right < n and right < left:
        backtrack(res, s+")", n, left, right+1)


# left表示左括号的数量, right表示右括号的数量
def generate_parenthesis(n):
    res = []
    s = ""
    backtrack(res, s, n, 0, 0)
    return res


if __name__ == "__main__":
    n = 3
    res = generate_parenthesis(n)
    print(res)
